docs/design/specs/runtime-async.md

```diff
@@ -21,30 +21,39 @@ Applicability of `MethodImplOptions.Async`:
 * Applying `MethodImplOptions.Async` to methods with `byref` or `ref-like` parameters is invalid.
 * Applying `MethodImplOptions.Async` to vararg methods is invalid.
 
+_[Note: these rules operate before generic substitution, meaning that a method which only meets requirements after substitution would not be considered as valid.]_
+
 Sync methods are all other methods.
 
 Unlike sync methods, async methods support suspension. Suspension allows async methods to yield control flow back to their caller at certain well-defined suspension points, and resume execution of the remaining method at a later time or location, potentially on another thread.
 
 Async methods also do not have matching return type conventions as sync methods. For sync methods, the stack should contain a value convertible to the stated return type before the `ret` instruction. For async methods, the stack should be empty in the case of `Task` or `ValueTask`, or the type argument in the case of `Task<T>` or `ValueTask<T>`.
 
-Async methods support the following suspension points:
+Async methods support suspension using one of the following methods:
 
-* Calling another method through the secondary encoding described in [### I.8.6.1.5 Method signatures]. No special instructions need to be provided. If the callee suspends, the caller will suspend as well.
-* Using new .NET runtime APIs to "await" an "INotifyCompletion" type. The signatures of these methods shall be:
   ```C#
   namespace System.Runtime.CompilerServices
   {
       public static class RuntimeHelpers
       {
           [MethodImpl(MethodImplOptions.Async)]
-          public static Task AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion { ... }
+          public static void AwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : INotifyCompletion { ... }
+          [MethodImpl(MethodImplOptions.Async)]
+          public static void UnsafeAwaitAwaiterFromRuntimeAsync<TAwaiter>(TAwaiter awaiter) where TAwaiter : ICriticalNotifyCompletion
```

Comment:
I assume the purpose of public `AwaitAwaiterFromRuntimeAsync` would be just for custom awaitables?

Previously, in async2 methods, Roslyn would generate a thunk call in cases when the await target has Task/ValueTask type **and is a method invocation**.  
Otherwise it would emit
```cs
                     {
                       var awaiter = arg.GetAwaiter();
                       if (awaiter.IsComplete())
                       {
                           UnsafeAwaitAwaiterFromRuntimeAsync(awaiter)
                       }
                       awaiter.GetResult()
                     }
```

Now awaiting anything that has Task type can be lowered into `Await` helper call. 
I.E.   `await obj.taskField` ==> `Await(obj.taskField)`